### PR TITLE
Always update endpoints in xDS cache

### PR DESCRIPTION
**Pull request by marcogschmidt**
When the translation of the Gloo API snapshot to the xDS one returns an error we currently skip updating the xDS cache. Gloo attempts to remove errored upstreams without compromising the consistency of the snapshot, but most errors still result in the xDS cache not being updated. By not updating the cache, Envoy will continue using the previous snapshot, which we assume to be valid. 

The problem with this approach is that endpoints are relatively ephemeral entities (think Kubernetes pods) and the snapshot that Gloo is using might become stale if any endpoint changes. 

With this fix Gloo will try to update the EDS information in the xDS cache even if an error occurred during translation. It does so by retrieving the previous xDS snapshot, updating only the EDS information, verifying the consistency of the resulting snapshot, and, if consistent, persisting it in the cache.
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/657

### CODE REVIEW

File: projects/gloo/pkg/xds/envoy_snapshot.go [hunk 1]
```diff
@@ -45,7 +46,7 @@ func NewSnapshot(version string,
 	endpoints []cache.Resource,
 	clusters []cache.Resource,
 	routes []cache.Resource,
-	listeners []cache.Resource) *EnvoySnapshot {
+	listeners []cache.Resource) cache.Snapshot {
```
Comment (yuval-k) [inline projects/gloo/pkg/xds/envoy_snapshot.go]: why change the return value?
IIRC the go best practice is "accept interfaces, return structs"

File: projects/gloo/pkg/xds/envoy_snapshot.go [hunk 2]
```diff
@@ -109,3 +110,41 @@ func (s *EnvoySnapshot) GetResources(typ string) cache.Resources {
...
+
+	return snapshotClone
+}
+
+func cloneItems(items map[string]cache.Resource) map[string]cache.Resource {
+	clonedItems := make(map[string]cache.Resource, len(items))
+	for k, v := range items {
+		resProto := v.ResourceProto()
+		// NOTE(marco): we have to use `github.com/golang/protobuf/proto.Clone()` to clone here,
+		// `github.com/golang/protobuf/proto.Clone()` will panic!
```
Comment (yuval-k) [inline projects/gloo/pkg/xds/envoy_snapshot.go]: did you mean gogo's version here?

File: projects/gloo/pkg/xds/envoy_snapshot.go (hunk 1 above)
Comment (marcogschmidt) [reply projects/gloo/pkg/xds/envoy_snapshot.go]: Did not know about that best practice. Having just read about it, it makes sense. I'll change it back.

File: projects/gloo/pkg/xds/envoy_snapshot.go (hunk 2 above)
Comment (marcogschmidt) [reply projects/gloo/pkg/xds/envoy_snapshot.go]: Yeah, will fix

**Pull request merged**